addons/im_livechat/static/src/legacy/public_livechat_chatbot.js

```diff
@@ -0,0 +1,874 @@
+                    // special case where we already have a "is typing" message displayed
+                    // can happen when the previous step did not trigger any message posted from the bot
+                    // e.g: previous step was "forward_operator" and no-one is available
+                    // -> in that case, don't wait and trigger the next step immediately
+                    nextStepDelay = 0;
+                } else {
+                    this._chatbotSetIsTyping();
+                }
+
+                this.nextStepTimeout = setTimeout(
+                    this._chatbotTriggerNextStep.bind(this), nextStepDelay);
+            }
+        }
+
+        if (!this._chatbotIsDisplayRestartButton()) {
+            this._chatWindow.$('.o_livechat_chatbot_main_restart').addClass('d-none');
+        }
+     },
+     /**
+      * See '_chatbotSaveSession'.
+      *
+      * We retrieve the livechat uuid from the session cookie since the livechat Widget is not yet
+      * initialized when we restore the chatbot state.
+      *
+      * We also clear any older keys that store a previously saved chatbot session.
+      * (In that case we clear the actual browser's local storage, we don't use the localStorage
+      * object as it does not allow browsing existing keys, see 'local_storage.js'.)
+      *
+      * @private
+      */
+    _chatbotRestoreSession: function (sessionCookie) {
+        const browserLocalStorage = window.localStorage;
+        if (browserLocalStorage && browserLocalStorage.length) {
+            for (let i = 0; i < browserLocalStorage.length; i++) {
+                const key = browserLocalStorage.key(i);
+                if (key.startsWith('im_livechat.chatbot.state.uuid_') &&
+                    key !== 'im_livechat.chatbot.state.uuid_' + JSON.parse(sessionCookie).uuid) {
+                        browserLocalStorage.removeItem(key);
+                }
+            }
+
+        }
+
+        let chatbotState = localStorage.getItem(
+            'im_livechat.chatbot.state.uuid_' + JSON.parse(sessionCookie).uuid);
+
+        if (chatbotState) {
+            chatbotState = JSON.parse(chatbotState);
+            this._isChatbot = true;
+            this._chatbot = chatbotState._chatbot;
+            this._chatbotCurrentStep = chatbotState._chatbotCurrentStep;
+        }
+     },
+    /**
+     * Register current chatbot step state into localStorage to be able to resume if the visitor
+     * goes to another website page or if he refreshes his page.
+     *
+     * (Will not work if the visitor switches browser but his livechat session will not be restored
+     *  anyway in that case, since it's stored into a cookie).
+     *
+     * @private
+     */
+    _chatbotSaveSession: function () {
+        const chatUuid = this._livechat.toData().uuid;
+        localStorage.setItem('im_livechat.chatbot.state.uuid_' + chatUuid, JSON.stringify({
```

Comment:
why not just store it in a cookie ?

do you need to store multiple session ? (if not the uuid do not need to be in the session key)

(if you need multiple session it's ok to store them in the localStorage imo)

Reply:
Cookies are the devil because they are sent through the HTTP requests, and I never need those information on the server side.

The trick with the uuid is used to avoid trying to later recover a session that is not tied to this conversation.
(e.g: you finish the conversation here and then go on another page that has another runbot configured, not likely but possible).

Reply:
:imp: